### DISCUSSION TITLE
2024-04-12 - Do you know how to write a great Pull Request?  - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by sethdaily** [SSWConsulting/SSW.Rules.Content#8351] [Chewing the Fat]
Hey SSWers,

Let's talk about these related rules...

main rule (where the comments are shown) - https://www.ssw.com.au/rules/write-a-good-pull-request/
https://www.ssw.com.au/rules/use-pull-request-templates-to-communicate-expectations/

What do you think about how you do PRs?
Do you have any PR tips?

**Comment by Freego1783** [comment]
PR templates are a good idea, I'd like to see them on every project I work on for consistancy! I feel like it's nice especially when you're a new developer on a project.

**Comment by tiagov8** [comment]
Dealing with GitHub as editor and reviewer in the past years made me naturally enhance my PRs and Issues to include the most information as possible. 

There is a huge gap between the knowledge (on the purpose of a PR) from the person who created a PR; and the reviewer who is 1st looking at it. That gap can be reduced with great titles and descriptions.

**Comment by tkapa** [comment]
PR templates are quite nice. 

A well structured PR with minimal change is ideal. When a change starts to become > 10 files there might be a problem.

**Comment by ncn-ssw** [comment]
The authoring of properly written PRs needs to be strictly enforced by team members to ensure that future team members understand why changes were made.

**Comment by ozdemir-mehmet-cer** [comment]
CER Templates are quite good. Who wrote them, someone from SSW?

**Comment by calumjs** [comment]
PR templates make sense, but I don't like the way ours begins with the tip before it gets to the actual change:

> Tip: Use SSW Rule Writer GPT for help with writing rules 🤖

When I get a notification about a PR I don't need to see a tip about using SSW Rule Writer GPT intended for the person who wrote the PR every time.
Maybe we could put it at the end? Or include instructions to delete it in the template?

**Reply by jaydenalchin** [reply]
I completely agree 🫡 https://github.com/SSWConsulting/SSW.Rules.Content/pull/8356

**Comment by AshAnil** [comment]
PR Templates are good to follow for consistency. It is also easier for the approver to understand why the PR was made.

**Comment by jakebayliss** [comment]
I like to add in bad code to see if Calum actually reviews my PR's. He always gets them @calumjs 
PR templates are a great idea - I think manually writing them just takes too long, especially if they are in good detail. I think there is a good in between with too much and too little

**Comment by chrisschultzssw** [comment]
I see a lot of pull requests that start with "**XS**✔ ◾ ".

While (I think) the meaning is self-explanatory, I don't see this in any of our PR rules (or in the links, e.g. https://gitmoji.dev/)...

**Comment by bahjatalaadelssw** [comment]
I liked the points in the video, but I use most of these techniques in file management hierarchies for footage and assets

**Comment by danielmackay** [comment]
Adding images or screenshots to PRs, helps to add helps more context for the reviewer.

Also, using it's super easy to review the entire branch for a PR in GitHub by pressing 'dot' on your keyboard.

![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/2636640/5851a331-fd33-441b-854b-f80e47e816a4)

Rider, has also just released GitHub PR integration so you can review and comment on all changes right inside of your favourite IDE 😎

![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/2636640/74b25946-88c5-4fc7-bccb-8b92030692c5)

**Reply by matt-goldman** [reply]
Absolutely agree. Before and after screenshots or even a done video are awesome in a PR.

**Reply by Dhruv-0987** [reply]
This is amazing, had no idea. I am never gonna review a PR on GitHub  or DevOps anymore

**Comment by GordonBeeming** [comment]
I love PR templates. If you don't feel aligned with your PR template for your project, you should talk to your team. Often its not great for everyone, but nobody wants to start the conversation and everyone is using it in anger

**Comment by cvionnet** [comment]
Having a template for PRs makes the process smoother and more consistent. This should be implemented for every project.

**Comment by YalalCakirSSW** [comment]
I like the idea of having templates for PRs. I think every team should have a template to achieve consistency. It would be great to see TimePro using one 👀

**Comment by matt-goldman** [comment]
PR templates are cool but there are other things we can focus on to help improve merge debt.

**Comment by stephanfakossw** [comment]
Nice and concise 😊

**Comment by ricksu978** [comment]
I like PR templates, it's like a gentle reminder for filling in the details.

**Comment by Calinator444** [comment]
I think there's a balance to be struck between writing detailed pull requests and writing descriptive pull requests. Often if I'm using git on my local machine i'll use `git log --oneline` to see what changes were recently made to a branch. Having a short descriptive message for each pull request is very useful for quickly identifying which changes were made.

I'd be curious if any other devs agree/disagree with this approach.

**Comment by wicksipedia** [comment]
There is certainly a lot of noise in our pr templates, they feel very over prescriptive.

My biggest pet peeve is when the titles for PRs don’t match what the code change was.

Also, I really dislike the term mob programming (which the template pushes), so much so that I strike it through. Was there really a mob? With pitchforks and burning torches angrily writing code?

I know what the term means but there are nicer ways to say you worked with someone else… why not stick with pair programming? The next best one I’ve come across is ensemble programming.

**Comment by steven0x51** [comment]
PR template is cool, but do we really need "2. What was changed"? As a good PR the changes should ideally be self-explanatory within the context of the code diffs

**Comment by tombui99** [comment]
PR template is a must. It makes the PR well structured and concise which makes it much easier for the reviewers.

**Comment by AntPolkanov** [comment]
I love our PR templates - they give some structure to PR descriptions.

I just cannot see the value of the 3rd question - `Did you do pair or mob programming (list names)?`. If feels a bit random... what value does it bring to the reviewer?

**Comment by amankumarrr** [comment]
PR templates are great for providing clear descriptions and titles, making the review process smoother. It's also crucial to keep PRs concise to minimize potential conflicts and make it easier to trace back to the original issue.

**Comment by TobyChurches** [comment]
I like the idea of PR templates! but I don't think its a 'one size fits all'. Each task will have different requirements, and should have different information in the description .

**Comment by tomek-i** [comment]
I would not call the GitHub PR template it a **great template**.

>Did you do pair or mob programming (list names)?

The section, like @AntPolkanov mentioned is @adamcogan specific :-D 


Our SugarLearning template has sections that are more valuable to the Reviewer as well as the Dev as the template serves also as a reminder to perform or consider certain aspects.


```md
<!-- Provide a general summary of your changes in the Title above -->

## Description
<!-- Describe your changes in detail -->

## Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
```

<hr/>

There are a few things we can do to improve the PR experience:
1. Using AI to do automated code reviews (static code analysis like deep source, snyk etc.)
2. use hooks to apply consistent formatting on PR creation and or other cool shenanigans (commit hooks, commit templates, husky)

And most Importantly:

3. If you forgot to check what you are actually staging andcommiting, create a draft or check the PR before submitting and fix anything you can find before submitting the PR (like your console.logs 🤣 )

**Comment by Harry-Ross** [comment]
I think PR templates are super important for consistency between PRs and improving the overall quality of PR descriptions. These rules are super important resources in ensuring that PR quality is maintained across all projects.

**Comment by Dhruv-0987** [comment]
PRs are important to maintain the history, to make sure that no bad code has been merged anywhere. I try to review PR with the author in a call, nothing can give better context than that

**Comment by jeoffreyfischer** [comment]
PR templates make it so much easier to write PRs. Sometimes we might forget some information e.g. context, screenshot, etc. 
And it's also important to remember that commit messages should be descriptive

**Comment by drwharris** [comment]
I love PR templates. So many little things we forget to do when we are in a rush that can be added as a reminder. Things like reviewing your own PR, ensuring you have made sure all existing tests pass, that you have added new tests.

**Comment by christoment** [comment]
I like PR templates - it acts as a reminder as well for people to do some checklist when submitting PR :)
Having a good PR description helps reviewing it a lot.

I also like the PR process, since it helps both reviewer and the author to both learn how other people think and see different practices in building software :)

**Comment by NadeekaJK** [comment]
PR templates are a great way to remind what info is expected in the description, and a great means to enforce quality of the code that you're trying to merge. (Eg: having tick boxes like "I've added test cases", "I've run existing unit tests")

**Comment by jimmy1992abc** [comment]
I like the PR templates. 
Is it possible to make the template in a different colour or font, so when people add their answers to the questions, they are distinctive for both the reviewer and contributor.

**Comment by Marxoz** [comment]
The better a PR is written the better chance that someone will review it. And at the end of the day, we want our changes to go in asap.

Adding automated T-Shirt sizing to the PR is a great way of informing devs how big their PR is, and sort of forces them to be mindful of their PR sizes

**Comment by SamBWagner** [comment]
In my experience. You should do your utmost to make your PR's palatable for the reviewers. I'll do 3-16 PR's per PBI just to make sure that I'm working in small, manageable chunks that are easy to review.

**Comment by duanxinhuan** [comment]
PR templates is cool but I think we can do things much cooler on GitHub

**Comment by MattParkerDev** [comment]
We use PR templates on the client I am on, and they are great to ensure I have filled out all the necessary details on the PR.

**Comment by jimmidier** [comment]
PR template is a must for code review efficiency and consistency

**Comment by jerryluossw** [comment]
Great rules, I believe the PRtemplate is very useful.
However, I wonder if we need **1. What triggered this change? (PBI link, Email Subject, conversation + reason, etc)**, because PR is  associated with PBI or Work Item in Azure DevOps, the information should be include in PBI or Work Item.

**Comment by jernejk** [comment]
I think the CTF was great! :)

While I like having PR templates, the "1. What triggered this change? (PBI link, Email Subject, conversation + reason, etc)" suggest that it's expected to not follow Scrum rules and PR needs to reference untrackable tasks. Ideally, PR on it's own is already associated to the correct GitHub issue, making this question irrelevant.

"2. What was changed?" is ok. I think "2. What is the impact?" is much better as it explicitly states what would be most useful for a PR review. While it's easy to see what changed (just look at changed files), it's difficult to determine how this PR impacts the code base. Do we have a new feature, does the bug fix add a new validation, will the user see a new input on the UI?

In short, templates are OK, but I think it would be nicer for them to be a bit more explicit in what we want and remove first question if it's not relevant. In Azure DevOps, we can't complete a PR without linking it to a PBI or bug which directly references what triggered the change.

**Comment by Aibono1225** [comment]
I like the PR template on Rules.Content, it always gives me useful guidance when I create the PR.

**Comment by bettybondoc** [comment]
I've been taught to make my commits small and my descriptions long. Well, not literally long but detailed - by providing enough context for my team and my future self. Six months down the line, even I might not remember the specifics of the changes I made so this approach ensures that when we look back at the history, we'll have a clear understanding of what was done and why.

**Comment by tino-liu** [comment]
All well written PR description gives the reviewer context before they look at the code. I want to first understand what is being done before I look into the micro details on how it is implemented. Great rule!

**Comment by PennyWalker** [comment]
Templates are great. The way I often use GitHub doesn't really make the template relevant to me because it's a basic upload of an image, but I try to put as much of the template in that makes sense

**Comment by SofieHong** [comment]
The CTF and the PR template was great 🌟 Especially for enhancing transparency and consistency

**Comment by zl24622** [comment]
Great CTF this week. I learned the PR templates and liked them.  The tips from the video are also helpful.

**Comment by leahy268** [comment]
The PR Template is good.  The video was awful though.  I mean I agreed with the content but it took a long time to get there.

**Comment by awaltos** [comment]
PR templates are a great tool. They are meant to make PRs faster to raise, more meaningful and consistent. It's important to make sure that the template is created with those ideals in mind.

**Comment by vladislav-kir** [comment]
PR templates are a great idea, since it might force the people, who would not normally do it, write a good description.

I did not like the video though, it could have been significantly shorter without loosing any important content....

**Comment by JackDevAU** [comment]
PR templates are cool! They help give a lot more context. 

![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/57518417/4e4ad58e-8c13-487f-abe1-9ec0b07c072d)
Figure: A Famous programmer once said. LGTM

**Comment by babakamyljanovssw** [comment]
I like PR templates, it makes the process consistent among team members

**Comment by MarcusIrmscherSSW** [comment]
My PR's are usually fairly simple but its still helpful having a template to start with.

**Comment by ozair-ashfaque** [comment]
PR templates are essential, they keep consistency and help the reviewer to give the context for the changes.
